### DISCUSSION TITLE
New Category whilst in Split Transaction can this be fixed? (Is this by design?)

### DISCUSSION THREAD

**Discussion by SJK55** [moneymanagerex/moneymanagerex#5132] [General]
Not a fault as such but, while in a Split Transaction, adding new sub categories and then the combobox accepts it, but if copy and paste is done for a second entry (in that same transaction) of that same new sub category, MMEX error message is Invalid Category. It's like it hasn't processed the new entry (sub category) even though it accepted it in the first entry.
I get around this by picking something existing, hitting Okay then going back and correcting it. (double the entry effort)

![Category not updating MMEX](https://user-images.githubusercontent.com/14268231/192198307-752916ec-b267-45e8-8d27-27967fd981b9.jpg)

**Comment by whalley** [comment]
I'll have a look into that.

**Reply by SJK55** [reply]
Thanks Mark.

**Comment by whalley** [comment]
I can't reproduce the issue with the latest 1.6.0Beta, tested on MacOS and Windows.

**Comment by SJK55** [comment]
Version: 1.5.22-Beta.1 64-bit is what I am using.
It's hard to reproduce. 
As I tried one way then another. (One way worked the other didn't)
I've tried to record it to show you the problem as I see it. If that is better.
These 2 MP4 files do show the problem.









https://user-images.githubusercontent.com/14268231/193705758-b7a0c358-c3da-4ab2-a265-816491b00e12.mp4


https://user-images.githubusercontent.com/14268231/193705775-be49c7e9-8bee-48c8-b12d-1e41ca71fab9.mp4

**Comment by vomikan** [comment]
It seems the autocomlete of mmcombobox does not updated on a new category added.
Bug

**Comment by whalley** [comment]
Bug report created https://github.com/moneymanagerex/moneymanagerex/issues/5158

**Reply by SJK55** [reply]
Thanks for understanding the issue I experienced. 
Have a great day!

**Comment by whalley** [comment]
Fix included in v1.6.0

**Reply by SJK55** [reply]
I'm not sure if any other users have had issues since this release but the split transactions seems slow, clunky and sometimes the amount of the total transaction seems not to update as you are adding more splits. Am I the only one having this issue? V1.6.0.

**Reply by whalley** [reply]
There are a few bugs that are fixed in v1.6.1
https://github.com/moneymanagerex/moneymanagerex/issues/5250
https://github.com/moneymanagerex/moneymanagerex/issues/5273

**Comment by tactilis** [comment]
@SJK55 Please try this:

1. Select the category for the split
2. Press Tab key to move to the Amount field
3. Type the amount of the split
4. Press the Return (Enter) key to:
  a) update the Total
  b) add a new row and move to the Category field ready for entering the next split
  
Does that resolve the slowness, clunkiness, and non-updating of the total you have reported?

**Reply by SJK55** [reply]
https://user-images.githubusercontent.com/14268231/201547600-d405ba45-cd72-4a11-a98e-7bd5ad342f68.mp4

Hi sorry for delay but couldn't get back to this until today. Enter at Amount does seem to update but the overall slowness of that box was never like that in the releases prior. I've attached a video to attempt to show what I am saying. But it has done other strange things such as blanking fields previously filled in, shuffling of entries. (I couldn't duplicate those in this example unfortunately)
I noticed that for amount if you use Tab instead of Enter it doesn't update the total. (Maybe that was my issue?)

**Comment by tactilis** [comment]
@SJK55
> But it has done other strange things such as blanking fields previously filled in, shuffling of entries.

With regard to your video, it looks like you are demonstrating the known bugs in 1.6.0.

For example, at 0:29 and at 1:39 it is clear that you are tabbing out of the Amount field (focus moves to Notes) rather than pressing Return as I suggested above. In v1.6.0 there was a bug that did not update the Total on change of focus.

There were also bugs in 1.6.0 which affected the display of categories in the splits, which your video also shows.

These bugs should be resolved in the forthcoming 1.6.1 release.

If you wish, you can install a 1.6.1 Beta build which includes the fixes from here: [https://ci.appveyor.com/project/moneymanagerex/moneymanagerex/builds/45335097](https://ci.appveyor.com/project/moneymanagerex/moneymanagerex/builds/45335097)

- Click on the Job name (64 bit or 32 bit)
- Click on the Artifacts tab
- Download the installer or the portable version

Does that resolve the problems that you have reported?

---

Some suggestions to make your use of the dialog faster:
- Reducing the width of the Split Transaction dialog would reduce the amount of mouse movement you are having to do.

- Pressing Return in the Amount field instead of tab will avoid you having to move the mouse and click Add Split for each new row.

- Typing the initial letters of the category in the Category field and selecting from the filtered list would avoid you having to display the Organize Categories dialog every time.

**Comment by SJK55** [comment]
Here is the other issue I was trying to illustrate. But I didn't go this far last time I posted. I mean it works but it's dam clunky.

https://user-images.githubusercontent.com/14268231/201797183-cf535810-6100-4761-a49d-e57a1212e6c4.mp4

**Comment by tactilis** [comment]
> Here is the other issue I was trying to illustrate. 

Issue noted.

I've tested this on 1.6.1 Beta and there _is_ a problem when the number of splits entered results in the entire list not being visible (i.e. when the scroll bar is displayed) **_and_** the Add Split button is used instead of pressing Return in the Amount field.

The video shows:
1. Entry of some splits by pressing Return in the Amount field to finish the current entry and add a new row.
2. Entry of some splits using using the Add Split button to add a new row.
3. Removal of some splits using the Remove Split button.
 
https://user-images.githubusercontent.com/4060092/201884808-8d6d4479-5837-4bc8-80d1-045d90940793.mp4
 
1. Pressing the Return key results in the new row being added cleanly as expected.
2. Using the Add Spilt button results in a flickery scroll & refresh.
3. Using the Remove Spilt button results in a flickery scroll & refresh.

The refresh at step 2 is not as bad as is seen in @SJK55's video. Maybe I have a faster PC.

Note that the problem can easily be seen by just repeatedly clicking the Add or Remove Split buttons as shown below:
 
https://user-images.githubusercontent.com/4060092/201888684-78807b10-6566-41da-8179-cd1a5f22e933.mp4
 

@SJK55 Can I suggest that you update to 1.6.1 Beta which will deal with the bugs that are in 1.6.0 **_and_** that you press Return to enter the amount and add the next row.

There's no need to report anything more against 1.6.0.

**Comment by tactilis** [comment]
@SJK55

@whalley has fixed the flickery scroll & refresh.. You can use the latest 1.6.1 Beta build if you wish to use it ahead of the 1.6.1 release.

**Reply by SJK55** [reply]
Thanks I downloaded the 1.6.1 Beta 64 bit version. 
Regarding Martin comment re faster PC this is what I have.
Intel(R) Core(TM) i5-8400 CPU @ 2.80GHz   2.81 GHz running 64 bit Windows 11 Enterprise 22H2 (latest updates) with 16.0 GB RAM
(6 Core CPU)

Big improvement but still jumping around.
See my video. If you know the category and sub it's fast! If you choose to hit enter and go to the secondary search it upsets the refresh split.
(for me anyway) I guess I'll accept it but with my PC specs I would have expected better performance. Maybe I'm pushing the database too hard?

https://user-images.githubusercontent.com/14268231/202054905-57509727-cd43-4419-962f-ad6f570338ee.mp4

**Comment by whalley** [comment]
OK, this is probably not how we envisaged people using the new category entry. Accessing the category dialog is quite inefficient and I can see how some extra processing is required each time it is called, as the active category comboboxes need fully need re-initialising as we don't currently know if a category was just selected or categories are added/changed/deleted. Will need some extra logic to check for this to avoid a refresh.

**Reply by SJK55** [reply]
Thanks.

**Comment by tactilis** [comment]
> This is probably not how we envisaged people using the new category entry.
> The active category comboboxes need fully need re-initialising as we don't currently know if a category was just selected or categories are added/changed/deleted

...and you do have a _very_ large number of categories @SJK55.

I doubt there are many people who create spending categories:subcategories to the level that you have with a different subcategory for each type of item:
 
![image](https://user-images.githubusercontent.com/4060092/202159857-e9d48f73-e5ec-4464-b6e8-114e167e8896.png)
 
![image](https://user-images.githubusercontent.com/4060092/202159917-981b46be-9468-4792-a6f5-83312e0fb971.png)
 
Seeing the different ways people make use of MMEX helps guide its development.

I'm interested to know what changes, if any, would allow you to more conveniently use the option of typing in the combobox to filter the categories, as you did for Alcohol in your video?
 
![image](https://user-images.githubusercontent.com/4060092/202148391-31fa62fa-d75f-4409-9e2e-62522f790f4d.png)

Any thoughts would be welcome.

---

Note that @n-stein has been doing some work that may be included in a later release, which would:

1. Allow categories to be nested to a greater extent than the current two category:subcategory levels.
Thus, you could have, for example:
```
Eating out
    Beverage
        Coffee
             All your types of coffee
             ...
             ...
        Tea
             All your types of tea
             ...
             ...
Household
    Battery
             All your types of batteries
             ...
             ...
````

2. In the Category combobox, permit you to type the initial letters of the category, then `:`  then the initial letters of a subcategory, and so on to more quickly filter the list

3. Make the Category combobox filter using the entered text anywhere in the category:subcategory:subcategory:...
So typing Brasil would list all of your types of Brasilia coffee  (ha! a _coffee filter_! 😉) 

---

To help us test MMEX developments, would it be possible to have a copy of your categories please?

You can export them by selecting `File | Export | QIF Files...` and then on this dialog:
 
![image](https://user-images.githubusercontent.com/4060092/202166374-44771935-7e52-4f6a-86da-43bf6ca58626.png)
 
- Tick Categories,
- Untick Accounts
- Click File and choose a file to export to

Then .zip the file and upload it here.

Is that OK with you please?

**Reply by SJK55** [reply]
Ok for development I'll do that. Nothing to worry about. Over the years I've  migrated from dbase - self created), spreadsheets to MS Money (2 versions) to the MoneyManager Ex (which is multi platform, whicj I picked it to access on android at the time) so I guess I've accumulated a lot of categories. I tend to push my apps and machines hard! PS this export took over half an hour or more! I have been trying to integrate a historical dbase database I had and it might have inflated the category count. I like the search combo box 'on return keying' when picking categories as I can find 'duplicate' sub categories and then tidy up on the go! 

[Categories gather over the years test data.zip](https://github.com/moneymanagerex/moneymanagerex/files/10026350/Categories.gather.over.the.years.test.data.zip)

**Reply by tactilis** [reply]
Wow, over 3,800 categories!

May I ask please:

1. Whether you have actually used them all (I'm guessing that this is the case), or are they just gathered from standard lists of item types (like the many types or batteries that you list)?

2. What the advantage is to you of having transactions categorised down to the individual item type (e.g battery type, style of coffee, blend of wine, etc) compared to having a aggregated categories (e.g. Batteries, or Coffee, Wine)?

3. If you do use all or most of these categories, then are there any other areas of MMEX that do not cope well? Categorised reports, for example?

<br>

The work that @n-stein is doing on nested categories may be helpful to you in the future if a script could be provided to convert your subcategories which use 'space hyphen space' into nested categories.  

Thank you for the data file. It's helpful.

**Reply by SJK55** [reply]
1 Martin, regarding the categories. Over the years I've used them and part of the fantastic super search (Enter on boxes in MoneyManager EX) is I can locate duplicates hidden in nooks and crannies. (and fix / delete them)
As I stated I've attempted to bring older databases info from Dbase (not to any standard fields unfortunately) MSMoney etc.
The batteries (LOL) is because I was fixing children's toys, parents in law watches etc. Easy to track later! 

2  Probably less so individual sub cats now with the notes helping, but I've got used to it. 

3 TBH I haven't used the reports that much.

4. Yes the idea of nested categories sounds great! I'll wait till I see the working model.

No problems!

**Comment by whalley** [comment]
@SJK55 Give this a go and let me know if it improves things for you....
https://ci.appveyor.com/project/moneymanagerex/moneymanagerex/builds/45404221/job/v8y46kteda48ml21/artifacts

**Reply by SJK55** [reply]
I just downloaded the latest beta from #5132 above and whatever you've done works a charm!
Smooth, responsive and what it should be doing. Thanks again for your patience.

**Reply by whalley** [reply]
If whilst you are in the category dialog and add/delete/rename a category then on returning back to the transaction dialog you will get the sluggish behaviour on return, but just selecting a category will be fine now.
